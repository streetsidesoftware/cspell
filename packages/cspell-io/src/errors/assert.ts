import { AssertionError } from './errors.js';

export function assert(value: unknown, message?: string): asserts value {
    if (!value) {
        throw new AssertionError(message ?? 'Assertion failed');
    }
}
