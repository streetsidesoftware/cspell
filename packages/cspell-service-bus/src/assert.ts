export function assert(value: unknown, message?: string | Error): asserts value {
    if (!value) {
        const err = message instanceof Error ? message : Error(message ?? 'AssertionError');
        throw err;
    }
}
