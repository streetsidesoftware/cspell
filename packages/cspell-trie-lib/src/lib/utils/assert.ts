export function assert(condition: unknown, message = 'Assert Failed'): asserts condition {
    if (condition) return;
    throw new Error(message);
}
