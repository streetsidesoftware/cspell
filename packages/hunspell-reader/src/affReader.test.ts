import assert from 'assert';
import * as path from 'path';

import * as aff from './affReader';
import { testing } from './affReader';

const dictionaryPath = path.resolve(__dirname, '../dictionaries');

function dictPath(name: string) {
    return path.resolve(dictionaryPath, name);
}

describe('parse an aff file', () => {
    const filename = dictPath('nl.aff');

    it('reads an aff file', async () => {
        await expect(aff.parseAffFile(filename)).resolves.toBeDefined();
    });

    it('tests parseAffixRule', () => {
        // cspell:ignore Ăśztetett
        const rule1 = testing.parseAffixRule({
            option: 'SFX',
            value: 'í 0 Ăśztetett/1115 [^aĂĄeĂŠiĂ­oĂłĂśĹuĂşĂźĹą] 20983',
        });
        expect(rule1).toBeDefined();
        assert(rule1);
        expect(rule1.type).toBe('SFX');
        expect(rule1.stripping).toBe('0');
        expect(rule1.affix).toBe('Ăśztetett/1115');
        expect(rule1.flag).toBe('í');
        expect(rule1.replace.source).toBe('$');
        expect(rule1.condition.source).toBe('[^aĂĄeĂŠiĂ­oĂłĂśĹuĂşĂźĹą]$');

        expect(testing.parseAffixRule({ option: 'SFX', value: '' })).toBeUndefined();
        expect(testing.parseAffixRule({ option: 'SFX', value: 'í' })).toBeUndefined();
        expect(testing.parseAffixRule({ option: 'SFX', value: 'í 0' })).toBeUndefined();

        const rule2 = testing.parseAffixRule({ option: 'SFX', value: 'í 0 Ăśztetett/1115' });
        assert(rule2);
        expect(rule2.type).toBe('SFX');
        expect(rule2.stripping).toBe('0');
        expect(rule2.affix).toBe('Ăśztetett/1115');
        expect(rule2.flag).toBe('í');
        expect(rule2.replace.source).toBe('$');
        expect(rule2.condition.source).toBe('.$');
    });

    it('tests tablePfxOrSfx with bad rule', () => {
        const x = testing.tablePfxOrSfx(undefined, testing.parseLine('SFX í Y 267'));
        const n = x.size;
        const y = testing.tablePfxOrSfx(x, testing.parseLine('SFX í'));
        expect(x.size).toBe(n);
        expect(y).toBe(x);
        const z = testing.tablePfxOrSfx(y, testing.parseLine('SFX í Ăś Ĺz/1109 Ăś 20974'));
        expect(z.size).toBe(1);
    });
});

describe('Validate parsing aff files', () => {
    test.each`
        affFile
        ${'en_US.aff'}
    `('Read aff $affFile', async ({ affFile }) => {
        const affFileName = dictPath(affFile);
        const affInfo = await aff.parseAffFile(affFileName);
        expect(affInfo).toMatchSnapshot();
    });
});
