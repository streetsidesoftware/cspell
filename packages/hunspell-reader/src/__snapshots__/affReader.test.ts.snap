// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate parsing aff files Read aff en_US.aff 1`] = `
{
  "COMPOUNDMIN": 1,
  "COMPOUNDRULE": [
    "n*1t",
    "n*mp",
  ],
  "ICONV": [
    {
      "from": "â€™",
      "to": "'",
    },
  ],
  "NOSUGGEST": "!",
  "ONLYINCOMPOUND": "c",
  "PFX": Map {
    "A" => {
      "combinable": true,
      "count": "1",
      "extra": [],
      "id": "A",
      "substitutionSets": Map {
        "^." => {
          "match": /\\^\\./,
          "substitutions": [
            {
              "attach": "re",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\^/,
            },
          ],
        },
      },
      "type": "PFX",
    },
    "I" => {
      "combinable": true,
      "count": "1",
      "extra": [],
      "id": "I",
      "substitutionSets": Map {
        "^." => {
          "match": /\\^\\./,
          "substitutions": [
            {
              "attach": "in",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\^/,
            },
          ],
        },
      },
      "type": "PFX",
    },
    "U" => {
      "combinable": true,
      "count": "1",
      "extra": [],
      "id": "U",
      "substitutionSets": Map {
        "^." => {
          "match": /\\^\\./,
          "substitutions": [
            {
              "attach": "un",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\^/,
            },
          ],
        },
      },
      "type": "PFX",
    },
    "C" => {
      "combinable": true,
      "count": "1",
      "extra": [],
      "id": "C",
      "substitutionSets": Map {
        "^." => {
          "match": /\\^\\./,
          "substitutions": [
            {
              "attach": "de",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\^/,
            },
          ],
        },
      },
      "type": "PFX",
    },
    "E" => {
      "combinable": true,
      "count": "1",
      "extra": [],
      "id": "E",
      "substitutionSets": Map {
        "^." => {
          "match": /\\^\\./,
          "substitutions": [
            {
              "attach": "dis",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\^/,
            },
          ],
        },
      },
      "type": "PFX",
    },
    "F" => {
      "combinable": true,
      "count": "1",
      "extra": [],
      "id": "F",
      "substitutionSets": Map {
        "^." => {
          "match": /\\^\\./,
          "substitutions": [
            {
              "attach": "con",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\^/,
            },
          ],
        },
      },
      "type": "PFX",
    },
    "K" => {
      "combinable": true,
      "count": "1",
      "extra": [],
      "id": "K",
      "substitutionSets": Map {
        "^." => {
          "match": /\\^\\./,
          "substitutions": [
            {
              "attach": "pro",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\^/,
            },
          ],
        },
      },
      "type": "PFX",
    },
  },
  "REP": [
    {
      "match": "a",
      "replaceWith": "ei",
    },
    {
      "match": "ei",
      "replaceWith": "a",
    },
    {
      "match": "a",
      "replaceWith": "ey",
    },
    {
      "match": "ey",
      "replaceWith": "a",
    },
    {
      "match": "ai",
      "replaceWith": "ie",
    },
    {
      "match": "ie",
      "replaceWith": "ai",
    },
    {
      "match": "alot",
      "replaceWith": "a_lot",
    },
    {
      "match": "are",
      "replaceWith": "air",
    },
    {
      "match": "are",
      "replaceWith": "ear",
    },
    {
      "match": "are",
      "replaceWith": "eir",
    },
    {
      "match": "air",
      "replaceWith": "are",
    },
    {
      "match": "air",
      "replaceWith": "ere",
    },
    {
      "match": "ere",
      "replaceWith": "air",
    },
    {
      "match": "ere",
      "replaceWith": "ear",
    },
    {
      "match": "ere",
      "replaceWith": "eir",
    },
    {
      "match": "ear",
      "replaceWith": "are",
    },
    {
      "match": "ear",
      "replaceWith": "air",
    },
    {
      "match": "ear",
      "replaceWith": "ere",
    },
    {
      "match": "eir",
      "replaceWith": "are",
    },
    {
      "match": "eir",
      "replaceWith": "ere",
    },
    {
      "match": "ch",
      "replaceWith": "te",
    },
    {
      "match": "te",
      "replaceWith": "ch",
    },
    {
      "match": "ch",
      "replaceWith": "ti",
    },
    {
      "match": "ti",
      "replaceWith": "ch",
    },
    {
      "match": "ch",
      "replaceWith": "tu",
    },
    {
      "match": "tu",
      "replaceWith": "ch",
    },
    {
      "match": "ch",
      "replaceWith": "s",
    },
    {
      "match": "s",
      "replaceWith": "ch",
    },
    {
      "match": "ch",
      "replaceWith": "k",
    },
    {
      "match": "k",
      "replaceWith": "ch",
    },
    {
      "match": "f",
      "replaceWith": "ph",
    },
    {
      "match": "ph",
      "replaceWith": "f",
    },
    {
      "match": "gh",
      "replaceWith": "f",
    },
    {
      "match": "f",
      "replaceWith": "gh",
    },
    {
      "match": "i",
      "replaceWith": "igh",
    },
    {
      "match": "igh",
      "replaceWith": "i",
    },
    {
      "match": "i",
      "replaceWith": "uy",
    },
    {
      "match": "uy",
      "replaceWith": "i",
    },
    {
      "match": "i",
      "replaceWith": "ee",
    },
    {
      "match": "ee",
      "replaceWith": "i",
    },
    {
      "match": "j",
      "replaceWith": "di",
    },
    {
      "match": "di",
      "replaceWith": "j",
    },
    {
      "match": "j",
      "replaceWith": "gg",
    },
    {
      "match": "gg",
      "replaceWith": "j",
    },
    {
      "match": "j",
      "replaceWith": "ge",
    },
    {
      "match": "ge",
      "replaceWith": "j",
    },
    {
      "match": "s",
      "replaceWith": "ti",
    },
    {
      "match": "ti",
      "replaceWith": "s",
    },
    {
      "match": "s",
      "replaceWith": "ci",
    },
    {
      "match": "ci",
      "replaceWith": "s",
    },
    {
      "match": "k",
      "replaceWith": "cc",
    },
    {
      "match": "cc",
      "replaceWith": "k",
    },
    {
      "match": "k",
      "replaceWith": "qu",
    },
    {
      "match": "qu",
      "replaceWith": "k",
    },
    {
      "match": "kw",
      "replaceWith": "qu",
    },
    {
      "match": "o",
      "replaceWith": "eau",
    },
    {
      "match": "eau",
      "replaceWith": "o",
    },
    {
      "match": "o",
      "replaceWith": "ew",
    },
    {
      "match": "ew",
      "replaceWith": "o",
    },
    {
      "match": "oo",
      "replaceWith": "ew",
    },
    {
      "match": "ew",
      "replaceWith": "oo",
    },
    {
      "match": "ew",
      "replaceWith": "ui",
    },
    {
      "match": "ui",
      "replaceWith": "ew",
    },
    {
      "match": "oo",
      "replaceWith": "ui",
    },
    {
      "match": "ui",
      "replaceWith": "oo",
    },
    {
      "match": "ew",
      "replaceWith": "u",
    },
    {
      "match": "u",
      "replaceWith": "ew",
    },
    {
      "match": "oo",
      "replaceWith": "u",
    },
    {
      "match": "u",
      "replaceWith": "oo",
    },
    {
      "match": "u",
      "replaceWith": "oe",
    },
    {
      "match": "oe",
      "replaceWith": "u",
    },
    {
      "match": "u",
      "replaceWith": "ieu",
    },
    {
      "match": "ieu",
      "replaceWith": "u",
    },
    {
      "match": "ue",
      "replaceWith": "ew",
    },
    {
      "match": "ew",
      "replaceWith": "ue",
    },
    {
      "match": "uff",
      "replaceWith": "ough",
    },
    {
      "match": "oo",
      "replaceWith": "ieu",
    },
    {
      "match": "ieu",
      "replaceWith": "oo",
    },
    {
      "match": "ier",
      "replaceWith": "ear",
    },
    {
      "match": "ear",
      "replaceWith": "ier",
    },
    {
      "match": "ear",
      "replaceWith": "air",
    },
    {
      "match": "air",
      "replaceWith": "ear",
    },
    {
      "match": "w",
      "replaceWith": "qu",
    },
    {
      "match": "qu",
      "replaceWith": "w",
    },
    {
      "match": "z",
      "replaceWith": "ss",
    },
    {
      "match": "ss",
      "replaceWith": "z",
    },
    {
      "match": "shun",
      "replaceWith": "tion",
    },
    {
      "match": "shun",
      "replaceWith": "sion",
    },
    {
      "match": "shun",
      "replaceWith": "cion",
    },
    {
      "match": "size",
      "replaceWith": "cise",
    },
  ],
  "SET": "UTF-8",
  "SFX": Map {
    "V" => {
      "combinable": false,
      "count": "2",
      "extra": [],
      "id": "V",
      "substitutionSets": Map {
        "e$" => {
          "match": /e\\$/,
          "substitutions": [
            {
              "attach": "ive",
              "attachRules": undefined,
              "extra": "",
              "remove": "e",
              "replace": /e\\$/,
            },
          ],
        },
        "[^e]$" => {
          "match": /\\[\\^e\\]\\$/,
          "substitutions": [
            {
              "attach": "ive",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
      },
      "type": "SFX",
    },
    "N" => {
      "combinable": true,
      "count": "3",
      "extra": [],
      "id": "N",
      "substitutionSets": Map {
        "e$" => {
          "match": /e\\$/,
          "substitutions": [
            {
              "attach": "ion",
              "attachRules": undefined,
              "extra": "",
              "remove": "e",
              "replace": /e\\$/,
            },
          ],
        },
        "y$" => {
          "match": /y\\$/,
          "substitutions": [
            {
              "attach": "ication",
              "attachRules": undefined,
              "extra": "",
              "remove": "y",
              "replace": /y\\$/,
            },
          ],
        },
        "[^ey]$" => {
          "match": /\\[\\^ey\\]\\$/,
          "substitutions": [
            {
              "attach": "en",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
      },
      "type": "SFX",
    },
    "X" => {
      "combinable": true,
      "count": "3",
      "extra": [],
      "id": "X",
      "substitutionSets": Map {
        "e$" => {
          "match": /e\\$/,
          "substitutions": [
            {
              "attach": "ions",
              "attachRules": undefined,
              "extra": "",
              "remove": "e",
              "replace": /e\\$/,
            },
          ],
        },
        "y$" => {
          "match": /y\\$/,
          "substitutions": [
            {
              "attach": "ications",
              "attachRules": undefined,
              "extra": "",
              "remove": "y",
              "replace": /y\\$/,
            },
          ],
        },
        "[^ey]$" => {
          "match": /\\[\\^ey\\]\\$/,
          "substitutions": [
            {
              "attach": "ens",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
      },
      "type": "SFX",
    },
    "H" => {
      "combinable": false,
      "count": "2",
      "extra": [],
      "id": "H",
      "substitutionSets": Map {
        "y$" => {
          "match": /y\\$/,
          "substitutions": [
            {
              "attach": "ieth",
              "attachRules": undefined,
              "extra": "",
              "remove": "y",
              "replace": /y\\$/,
            },
          ],
        },
        "[^y]$" => {
          "match": /\\[\\^y\\]\\$/,
          "substitutions": [
            {
              "attach": "th",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
      },
      "type": "SFX",
    },
    "Y" => {
      "combinable": true,
      "count": "1",
      "extra": [],
      "id": "Y",
      "substitutionSets": Map {
        ".$" => {
          "match": /\\.\\$/,
          "substitutions": [
            {
              "attach": "ly",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
      },
      "type": "SFX",
    },
    "G" => {
      "combinable": true,
      "count": "2",
      "extra": [],
      "id": "G",
      "substitutionSets": Map {
        "e$" => {
          "match": /e\\$/,
          "substitutions": [
            {
              "attach": "ing",
              "attachRules": undefined,
              "extra": "",
              "remove": "e",
              "replace": /e\\$/,
            },
          ],
        },
        "[^e]$" => {
          "match": /\\[\\^e\\]\\$/,
          "substitutions": [
            {
              "attach": "ing",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
      },
      "type": "SFX",
    },
    "J" => {
      "combinable": true,
      "count": "2",
      "extra": [],
      "id": "J",
      "substitutionSets": Map {
        "e$" => {
          "match": /e\\$/,
          "substitutions": [
            {
              "attach": "ings",
              "attachRules": undefined,
              "extra": "",
              "remove": "e",
              "replace": /e\\$/,
            },
          ],
        },
        "[^e]$" => {
          "match": /\\[\\^e\\]\\$/,
          "substitutions": [
            {
              "attach": "ings",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
      },
      "type": "SFX",
    },
    "D" => {
      "combinable": true,
      "count": "4",
      "extra": [],
      "id": "D",
      "substitutionSets": Map {
        "e$" => {
          "match": /e\\$/,
          "substitutions": [
            {
              "attach": "d",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
        "[^aeiou]y$" => {
          "match": /\\[\\^aeiou\\]y\\$/,
          "substitutions": [
            {
              "attach": "ied",
              "attachRules": undefined,
              "extra": "",
              "remove": "y",
              "replace": /y\\$/,
            },
          ],
        },
        "[^ey]$" => {
          "match": /\\[\\^ey\\]\\$/,
          "substitutions": [
            {
              "attach": "ed",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
        "[aeiou]y$" => {
          "match": /\\[aeiou\\]y\\$/,
          "substitutions": [
            {
              "attach": "ed",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
      },
      "type": "SFX",
    },
    "T" => {
      "combinable": false,
      "count": "4",
      "extra": [],
      "id": "T",
      "substitutionSets": Map {
        "e$" => {
          "match": /e\\$/,
          "substitutions": [
            {
              "attach": "st",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
        "[^aeiou]y$" => {
          "match": /\\[\\^aeiou\\]y\\$/,
          "substitutions": [
            {
              "attach": "iest",
              "attachRules": undefined,
              "extra": "",
              "remove": "y",
              "replace": /y\\$/,
            },
          ],
        },
        "[aeiou]y$" => {
          "match": /\\[aeiou\\]y\\$/,
          "substitutions": [
            {
              "attach": "est",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
        "[^ey]$" => {
          "match": /\\[\\^ey\\]\\$/,
          "substitutions": [
            {
              "attach": "est",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
      },
      "type": "SFX",
    },
    "R" => {
      "combinable": true,
      "count": "4",
      "extra": [],
      "id": "R",
      "substitutionSets": Map {
        "e$" => {
          "match": /e\\$/,
          "substitutions": [
            {
              "attach": "r",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
        "[^aeiou]y$" => {
          "match": /\\[\\^aeiou\\]y\\$/,
          "substitutions": [
            {
              "attach": "ier",
              "attachRules": undefined,
              "extra": "",
              "remove": "y",
              "replace": /y\\$/,
            },
          ],
        },
        "[aeiou]y$" => {
          "match": /\\[aeiou\\]y\\$/,
          "substitutions": [
            {
              "attach": "er",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
        "[^ey]$" => {
          "match": /\\[\\^ey\\]\\$/,
          "substitutions": [
            {
              "attach": "er",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
      },
      "type": "SFX",
    },
    "Z" => {
      "combinable": true,
      "count": "4",
      "extra": [],
      "id": "Z",
      "substitutionSets": Map {
        "e$" => {
          "match": /e\\$/,
          "substitutions": [
            {
              "attach": "rs",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
        "[^aeiou]y$" => {
          "match": /\\[\\^aeiou\\]y\\$/,
          "substitutions": [
            {
              "attach": "iers",
              "attachRules": undefined,
              "extra": "",
              "remove": "y",
              "replace": /y\\$/,
            },
          ],
        },
        "[aeiou]y$" => {
          "match": /\\[aeiou\\]y\\$/,
          "substitutions": [
            {
              "attach": "ers",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
        "[^ey]$" => {
          "match": /\\[\\^ey\\]\\$/,
          "substitutions": [
            {
              "attach": "ers",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
      },
      "type": "SFX",
    },
    "S" => {
      "combinable": true,
      "count": "4",
      "extra": [],
      "id": "S",
      "substitutionSets": Map {
        "[^aeiou]y$" => {
          "match": /\\[\\^aeiou\\]y\\$/,
          "substitutions": [
            {
              "attach": "ies",
              "attachRules": undefined,
              "extra": "",
              "remove": "y",
              "replace": /y\\$/,
            },
          ],
        },
        "[aeiou]y$" => {
          "match": /\\[aeiou\\]y\\$/,
          "substitutions": [
            {
              "attach": "s",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
        "[sxzh]$" => {
          "match": /\\[sxzh\\]\\$/,
          "substitutions": [
            {
              "attach": "es",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
        "[^sxzhy]$" => {
          "match": /\\[\\^sxzhy\\]\\$/,
          "substitutions": [
            {
              "attach": "s",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
      },
      "type": "SFX",
    },
    "P" => {
      "combinable": true,
      "count": "3",
      "extra": [],
      "id": "P",
      "substitutionSets": Map {
        "[^aeiou]y$" => {
          "match": /\\[\\^aeiou\\]y\\$/,
          "substitutions": [
            {
              "attach": "iness",
              "attachRules": undefined,
              "extra": "",
              "remove": "y",
              "replace": /y\\$/,
            },
          ],
        },
        "[aeiou]y$" => {
          "match": /\\[aeiou\\]y\\$/,
          "substitutions": [
            {
              "attach": "ness",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
        "[^y]$" => {
          "match": /\\[\\^y\\]\\$/,
          "substitutions": [
            {
              "attach": "ness",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
      },
      "type": "SFX",
    },
    "M" => {
      "combinable": true,
      "count": "1",
      "extra": [],
      "id": "M",
      "substitutionSets": Map {
        ".$" => {
          "match": /\\.\\$/,
          "substitutions": [
            {
              "attach": "'s",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
      },
      "type": "SFX",
    },
    "B" => {
      "combinable": true,
      "count": "3",
      "extra": [],
      "id": "B",
      "substitutionSets": Map {
        "[^aeiou]$" => {
          "match": /\\[\\^aeiou\\]\\$/,
          "substitutions": [
            {
              "attach": "able",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
        "ee$" => {
          "match": /ee\\$/,
          "substitutions": [
            {
              "attach": "able",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
        "[^aeiou]e$" => {
          "match": /\\[\\^aeiou\\]e\\$/,
          "substitutions": [
            {
              "attach": "able",
              "attachRules": undefined,
              "extra": "",
              "remove": "e",
              "replace": /e\\$/,
            },
          ],
        },
      },
      "type": "SFX",
    },
    "L" => {
      "combinable": true,
      "count": "1",
      "extra": [],
      "id": "L",
      "substitutionSets": Map {
        ".$" => {
          "match": /\\.\\$/,
          "substitutions": [
            {
              "attach": "ment",
              "attachRules": undefined,
              "extra": "",
              "remove": "0",
              "replace": /\\$/,
            },
          ],
        },
      },
      "type": "SFX",
    },
  },
  "TRY": "esianrtolcdugmphbyfvkwzESIANRTOLCDUGMPHBYFVKWZ'",
  "WORDCHARS": "0123456789",
}
`;
